Assets/Scripts/Infrastructure/NetworkObjectPool.cs

```diff
@@ -4,39 +4,37 @@
 using Unity.Netcode;
 using UnityEngine;
 using UnityEngine.Assertions;
+using UnityEngine.Pool;
 
 namespace Unity.BossRoom.Infrastructure
 {
     /// <summary>
-    /// Object Pool for networked objects, used for controlling how objects are spawned by Netcode. Netcode by default will allocate new memory when spawning new
-    /// objects. With this Networked Pool, we're using custom spawning to reuse objects.
+    /// Object Pool for networked objects, used for controlling how objects are spawned by Netcode. Netcode by default
+    /// will allocate new memory when spawning new objects. With this Networked Pool, we're using the ObjectPool to
+    /// reuse objects.
     /// Boss Room uses this for projectiles. In theory it should use this for imps too, but we wanted to show vanilla spawning vs pooled spawning.
-    /// Hooks to NetworkManager's prefab handler to intercept object spawning and do custom actions
+    /// Hooks to NetworkManager's prefab handler to intercept object spawning and do custom actions.
     /// </summary>
     public class NetworkObjectPool : NetworkBehaviour
     {
-        private static NetworkObjectPool _instance;
-
-        public static NetworkObjectPool Singleton { get { return _instance; } }
+        public static NetworkObjectPool Singleton { get; private set; }
 
         [SerializeField]
         List<PoolConfigObject> PooledPrefabsList;
 
-        HashSet<GameObject> prefabs = new HashSet<GameObject>();
-
-        Dictionary<GameObject, Queue<NetworkObject>> pooledObjects = new Dictionary<GameObject, Queue<NetworkObject>>();
+        HashSet<GameObject> m_Prefabs = new HashSet<GameObject>();
 
-        private bool m_HasInitialized = false;
+        Dictionary<GameObject, ObjectPool<NetworkObject>> m_PooledObjects = new Dictionary<GameObject, ObjectPool<NetworkObject>>();
 
         public void Awake()
         {
-            if (_instance != null && _instance != this)
+            if (Singleton != null && Singleton != this)
             {
-                Destroy(this.gameObject);
+                Destroy(gameObject);
             }
             else
             {
-                _instance = this;
+                Singleton = this;
```

Comment:
True, I'll remove the methods we are not using.
For "CreateInstance" I thought it was placed in a separate method for performance reasons with the attribute above it to inline it, but in this case inlining a one-line method should be the same as just calling that one line instead of the method, so I don't think this is actually helping. I'll remove it.